src/java.base/share/classes/sun/security/ssl/SSLSessionImpl.java

```diff
@@ -1456,6 +1494,267 @@ public List<SNIServerName> getRequestedServerNames() {
+             *         "exporter", Hash(context_value), key_length)
+             *
+             * Derive-Secret(Secret, Label, Messages) =
+             *     HKDF-Expand-Label(Secret, Label,
+             *         Transcript-Hash(Messages), Hash.length)
+             */
+
+            try {
+                // Use the ciphersuite's hashAlg for these calcs.
+                HashAlg hashAlg = cipherSuite.hashAlg;
+                KDF hkdf = KDF.getInstance(hashAlg.hkdfAlgorithm);
+
+                // First calculate the inner Derive-Secret(Secret, label, "")
+                MessageDigest md;
+                byte[] emptyHash;
+
+                // Create the "" digest...
+                try {
+                    md = MessageDigest.getInstance(hashAlg.toString());
+                    emptyHash = md.digest();
+                } catch (NoSuchAlgorithmException nsae) {
+                    throw new ProviderException(
+                            "Hash algorithm " + cipherSuite.hashAlg.name +
+                                    " is not available", nsae);
+                }
+
+                // ...then the hkdfInfo...
+                byte[] hkdfInfo = SSLSecretDerivation.createHkdfInfo(
+                        hkdfInfoLabel, emptyHash, hashAlg.hashLength);
+
+                // ...then the "inner" HKDF-Expand-Label() to get the
+                // derivedSecret that is used as the Secret in the "outer"
+                // HKDF-Expand-Label().
+                SecretKey derivedSecret = hkdf.deriveKey("TlsKey",
+                        HKDFParameterSpec.expandOnly(exporterMasterSecret,
+                        hkdfInfo, hashAlg.hashLength));
+                try {
+                    // Now do the "outer" HKDF-Expand-Label.
+                    //     HKDF-Expand-Label(derivedSecret, "exporter",
+                    //         Hash(context_value), key_length)
+
+                    // If a context was supplied, use it, otherwise, use the
+                    // previous hashed value of ""...
+                    byte[] hash = ((context.length > 0) ?
+                            md.digest(context) : emptyHash);
+
+                    // ...now the hkdfInfo...
+                    hkdfInfo = SSLSecretDerivation.createHkdfInfo(
+                            ("tls13 exporter").getBytes(StandardCharsets.UTF_8),
+                            hash, length);
+
+                    // ...now the final expand.
+                    return ((keyAlg != null) ?
+                            hkdf.deriveKey(keyAlg,
+                                    HKDFParameterSpec.expandOnly(derivedSecret,
+                                            hkdfInfo, length)) :
+                            hkdf.deriveData(
+                                    HKDFParameterSpec.expandOnly(derivedSecret,
+                                            hkdfInfo, length)));
+                } finally {
+                    KeyUtil.destroySecretKeys(derivedSecret);
+                }
```

Comment:
The `derivedSecret` is destroyed and cleared. How about `exporterMasterSecret`? Shall we also destroy it perhaps when the session is invalidated or closed?

Reply:
In RFC 8446/Section 7.1, the "derived" secret and "Master Secret" are temporary variables used to derive the application secrets, the exporter_master_secret, and the resumption_master_secret.  These variables are long-lived and don't need to be cleared like these two.

When we invalidate a session, that simply marks the session as unresumable, and doesn't destroy the session for active connections.  We don't "close" sessions.